src/v/cluster/shard_placement_table.cc

```diff
@@ -0,0 +1,451 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/shard_placement_table.h"
+
+#include "cluster/cluster_utils.h"
+#include "cluster/logger.h"
+#include "cluster/topic_table.h"
+
+namespace cluster {
+
+std::ostream&
+operator<<(std::ostream& o, shard_placement_table::hosted_status s) {
+    switch (s) {
+    case shard_placement_table::hosted_status::receiving:
+        return o << "receiving";
+    case shard_placement_table::hosted_status::hosted:
+        return o << "hosted";
+    case shard_placement_table::hosted_status::obsolete:
+        return o << "obsolete";
+    }
+    __builtin_unreachable();
+}
+
+std::ostream& operator<<(
+  std::ostream& o, const shard_placement_table::shard_local_state& ls) {
+    fmt::print(
+      o, "{{log_revision: {} status: {}}}", ls.log_revision, ls.status);
+    return o;
+}
+
+std::ostream&
+operator<<(std::ostream& o, const shard_placement_table::placement_state& gs) {
+    fmt::print(
+      o,
+      "{{local: {}, target: {}, shard_revision: {}, "
+      "is_initial_at_revision: {}, next: {}}}",
+      gs.local,
+      gs.target,
+      gs.shard_revision,
+      gs._is_initial_at_revision,
+      gs._next);
+    return o;
+}
+
+ss::future<> shard_placement_table::initialize(
+  const topic_table& topics, model::node_id self) {
+    // We expect topic_table to remain unchanged throughout the loop because the
+    // method is supposed to be called after local controller replay is finished
+    // but before we start getting new controller updates from the leader.
+    auto tt_version = topics.topics_map_revision();
+    for (const auto& [ns_tp, md_item] : topics.all_topics_metadata()) {
+        vassert(
+          tt_version == topics.topics_map_revision(),
+          "topic_table unexpectedly changed");
+        co_await ss::maybe_yield();
```

Comment:
nit: take a look at `ssx:async_for_each_coro` for effective yielding between the loops

Reply:
haha, new arcane knowledge just dropped :)